### PR TITLE
feat: intro univ3 and prep seeding scripting for ebtc/wbtc pool

### PR DESCRIPTION
**Pull request by petrovska-petro**
tackles #8 

```
brownie run scripts/univ3_management pool_creation_and_init_seeding
```

### CODE REVIEW

**Review by sajanrajdev — COMMENTED** [review]
## Review
- The UniV3 helpes and library were integrally imported from the badger-multisig [GreatApeSafe](https://github.com/Badger-Finance/badger-multisig/blob/main/great_ape_safe/ape_api/uni_v3.py). These include helpers to functions to create and manage liquidity positions, swap, collect fees and visualize the state of LPs.
- The required mainnet addresses, as well as the Vault and Trops addresses, were added to the address book.
- A draft for the eBTC pool creation and seeding was developed:
    - Pool is created using the 0.05% fee tier as planned
    - Pool is initialized for a 1:1 price 
    - Positions are minted for each one of the planned ranges using the percentages predefined by BALCO.
    - Used LIQ as a dummy token while eBTC is deployed to mainnet

@petrovska-petro Implementations LGTM, my only comment is that we should use this chance to add unit tests for at least the operations that we intend to use. It is ok if these are meant to be run against mainnet instead of Sepolia (as opposed to the existing ones). 

My concern relates to the issue in which, not too long ago, we almost execute a swap using this library that had a [misestimation](https://github.com/Badger-Finance/badger-multisig/pull/1464) of the min_out and could have resulted in drain of value. Considering how often we will use this library, I think we need higher certainty into its accuracy.

**Comment by sajanrajdev** [comment]
Tackles https://github.com/ebtc-protocol/ebtc-multisig/issues/11

**Comment by petrovska-petro** [comment]
great suggestion @sajanrajdev i added tests on the last commits for the actions intended for now

one thing im still researching is something to note on the `test_position_creation` that the ticks of the position when converting them from the upper and lower ticks does not match the intended. values are a bit off:

```
price_lower_ticket=0.9891
price_higher_tick=0.995
```

run: ` brownie test tests/univ3/* -s `

im still investigating if related to tick spacing

**Comment by petrovska-petro** [comment]
pushed 5241b91ba1edc62997a7f461694bc89c054014c2 which i think was a mistaken assumptions inherit from the fact that badgerwbtc and ethgtc shared the same fee tier, but this one suggested in the forum is meant to be 0.05% which by result the tick spacing will be different. now it should represent a more precise representation of the nft positions created given the price targets. will be great if you could also provide feedback if you see something im missing out @sajanrajdev

**Review by rayeaster — APPROVED** [review]
LGTM to have scripts to setup UniV3 LP with respect to BALCO decision

File: scripts/univ3_management.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    price_099 = 0.99
+    price_1 = 1
+    price_101 = 1.01
+    price_108 = 1.08
+
+    # @note these figures are note definitive, will be updated in the future!
+    token_0_init_seeding_amount = 2.5e8 if token0 == wbtc.address else 2.5e18
+    token_1_init_seeding_amount = 2.5e18 if token1 == liq.address else 2.5e8
+
+    # single-side lp only ebtc
```
Comment (rayeaster) [inline scripts/univ3_management.py:68]: do we assume `token0` is eBTC here?

File: scripts/univ3_management.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    )
+    safe.uni_v3.mint_position(
+        pool,
+        price_1,
+        price_101,
+        token_0_init_seeding_amount * PCT_40,
+        token_1_init_seeding_amount * PCT_40,
+    )
+
+    # single-side only wbtc
```
Comment (rayeaster) [inline scripts/univ3_management.py:89]: do we assume `token1` is `wBTC` here?

**Review by sajanrajdev — APPROVED** [review]
## Additional Review
- Introduces a test case for an pool creation using the expected specs for eBTC/wBTC (with LIQ as dummy token1 for now)
- Introduces a test case for the treasury's position creation for the 1 - 1.01 range
- Allows for tick spacing to be changed dynamically for pools of different fee tiers when minting new positions
- All tests are passing: 
image


LGTM

File: tests/univ3/test_position_creation.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        PRICE_1,
+        PRICE_101,
+        token_0_init_seeding_amount,
+        token_1_init_seeding_amount,
+    )
+
+    positions = treasury.uni_v3.nonfungible_position_manager.positions(token_id)
+
+    # TODO: rounding to 4 decimal places, values are not matching the expected?
+    price_lower_ticket = round(
```
Comment (sajanrajdev) [inline tests/univ3/test_position_creation.py:40]: Minor typo in here. Mean ticker or tick, right?

Comment (petrovska-petro) [reply tests/univ3/test_position_creation.py]: good catch, fixed in 9d360d763ed5c69b7e25fb393919e271bf74f4da

File: scripts/univ3_management.py (hunk 1 above)
Comment (petrovska-petro) [reply scripts/univ3_management.py:68]: this is still sort of a placeholder in light of non `ebtc` in prod. order of which one will be token0 or token1 will depending on the internally ordering of the each token address itself when deploying the pool 

https://github.com/Uniswap/v3-core/blob/main/contracts/UniswapV3Factory.sol#L41

File: scripts/univ3_management.py (hunk 2 above)
Comment (petrovska-petro) [reply scripts/univ3_management.py:89]: same as https://github.com/ebtc-protocol/ebtc-multisig/pull/18#discussion_r1489981513

File: scripts/cdp_management_lens.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    for cdp_id in cdps_safe_owned:
+        C.print(f"[green]Inspecting cdp id: {cdp_id}\n[/green]")
+        (
+            cdp_id_debt,
+            cdp_id_coll,
+            _,
+            _,
+            _,
+            _,
+        ) = safe.ebtc.cdp_manager.Cdps(cdp_id)
```
[the remark this answers is not in the document]
Comment (petrovska-petro) [reply scripts/cdp_management_lens.py]: suggested added in 390a7d92c3dd37f3d5d52806101cf6a2d707e0ea

**Review by sajanrajdev — APPROVED** [review]
LGTM

**Review by rayeaster — APPROVED** [review]

**Pull request merged**